x/rollup/tx/helpers/l1_data_decorator.go

```diff
@@ -0,0 +1,91 @@
+package helpers
+
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	rollupkeeper "github.com/polymerdao/monomer/x/rollup/keeper"
+)
+
+// ConsumeGasForL1DataDecorator will consume gas to compensate the sequencer
+// for posting the transaction to Ethereum. The gas cost is calculated based
+// on the Ecotone upgrade and the sequencer is expected to post the transaction
+// using blobs.
+type ConsumeGasForL1DataDecorator struct {
+	rollupKeeper *rollupkeeper.Keeper
+}
+
+func NewConsumeGasForL1DataDecorator(rollupKeeper *rollupkeeper.Keeper) ConsumeGasForL1DataDecorator {
+	return ConsumeGasForL1DataDecorator{
+		rollupKeeper: rollupKeeper,
+	}
+}
+
+func (d ConsumeGasForL1DataDecorator) AnteHandle(
+	ctx sdk.Context, //nolint:gocritic // hugeparam
+	tx sdk.Tx,
+	simulate bool,
+	next sdk.AnteHandler,
+) (sdk.Context, error) {
+	l1BlockInfo, err := d.rollupKeeper.GetL1BlockInfo(ctx)
+	if err != nil {
+		return ctx, fmt.Errorf("get l1 block info: %w", err)
+	}
+
+	baseFee := new(big.Int).SetBytes(l1BlockInfo.BaseFee)
+	baseFeeScalar := big.NewInt(int64(l1BlockInfo.BaseFeeScalar))
+	blobBaseFee := new(big.Int).SetBytes(l1BlockInfo.BlobBaseFee)
+	blobBaseFeeScalar := big.NewInt(int64(l1BlockInfo.BlobBaseFeeScalar))
+
+	l1GasUsed, err := getL1GasUsed(tx)
+	if err != nil {
+		return ctx, fmt.Errorf("get l1 gas used: %w", err)
+	}
+
+	const baseFeeMultiplier = 16
+	const divisor = 16e6
+
+	// scaledBaseFee calculation: scaledBaseFee = l1BaseFee * lBaseFeeScalar * 16
+	scaledBaseFee := new(big.Int).Mul(new(big.Int).Mul(baseFee, baseFeeScalar), big.NewInt(baseFeeMultiplier))
+
+	// scaledBlobBaseFee calculation: scaledBlobBaseFee = l1BlobBaseFee * l1BlobBaseFeeScalar
+	scaledBlobBaseFee := new(big.Int).Mul(blobBaseFee, blobBaseFeeScalar)
+
+	// l1DataCost calculation: l1DataCost = [l1GasUsed * (scaledBaseFee + scaledBlobBaseFee)] / 16e6
+	l1DataCost := new(big.Int).Div(new(big.Int).Mul(l1GasUsed, new(big.Int).Add(scaledBaseFee, scaledBlobBaseFee)), big.NewInt(divisor))
+
+	if !l1DataCost.IsUint64() {
+		return ctx, fmt.Errorf("l1 data cost overflow: %s", l1DataCost)
+	}
```

Comment:
is this a possibility or is it technically impossible?

Reply:
In practice I don't think that this block should ever be hit, however I wanted to add the extra safeguard since we're dealing with a lot of multiplication here. This would prevent against the case where one of the L1 attributes either unexpectedly increased or was misconfigured in the system deposit tx.

That being said, I'm not sure how useful this check is. Do you think I should just remove this block for the overflow case since it shouldn't ever be hit in production?